TechnicalDocuments/Proposals/BuildersToFunctions.md

```diff
@@ -0,0 +1,161 @@
+
+## Proposed solution
+
+**Note:** The proposed solution assumes that a way to authenticate HTTP requests is available at `Current` level (in this case something in the shape of `MainUserSession` abstracted via `UserSession`).
+
+Let's use `MedicalHistoryBuilder.swift` as an example, since it offers a trivial API: 
+
+``` swift
+init(session: UserSession, visuals: VisualDependenciesProtocol)
+```
+
+And the corresponding public API:
+
+```swift
+func make(modal: Flow) -> UIViewController
+```
+
+Given the Module where it's located it could be migrated to:
+
+```swift
+extension Builder {
+    enum ClinicalRecords {
+        func makeMedicalHistory(modal: Flow) -> UIViewController
+    }
+}
+```
+
+Particular medical histories (e.g. for minors) can also be created, but in this case, the minor session would be passed instead of relying on `Current`. This situation (with child builders) can be observed at `ClinicalRecordsChildBuilders.swift`:
+
+```swift
+extension Builder {
+    enum ClinicalRecords {
+        func makeMedicalHistory(with session: Session, modal: Flow) -> UIViewController
+        func makeMedicalHistory(modal: Flow) -> UIViewController
+    }
+}
+```
+
+This would remove the intricated structure we have with builders and its children.
+
+A side effect of this approach, because we are dealing with functions (versus structs) we are able to remove local state. An example of such case is `enableClinicalDownloads` at `ClinicalRecordsChildBuilders`. You can see the code smell here:
+
+```swift
+builders.appointmentList(with: session,
+                         enableClinicalDownloads: builders.enableClinicalDownloads,
+                         addressSwitches: builders.addressSwitches)
+```
+
+With the free functions approach, this would never be possible to do, since builders don't hold state:
+
+```swift
+extension Builder {
+    enum ClinicalRecords {
+        func makeAppointmentList(session: UserSession,
+                                 enableClinicalDownloads: Bool,
+                                 addressSwitches: AddressSwitchesProtocol,
+                                 primary: Flow,
+                                 secondary: Flow) -> UIViewController
+    }
+}
+```
+
+At the call site, we would replace this:
+
+```swift
+case let .appointmentList(session):
+    builders
+            .appointmentList(with: session,
+                             enableClinicalDownloads: builders.enableClinicalDownloads,
+                             addressSwitches: builders.addressSwitches)
+            .make(primary: navigation, secondary: modal)
+            |> navigation.present
+```
+
+With this:
+
+```swift
+case let .appointmentList(session):
+    Builder.ClinicalRecords
+            .makeAppointmentList(with: session,
+                                enableClinicalDownloads: enableClinicalDownloads,
+                                addressSwitches: addressSwitches,
```

Comment:
So would `addressSwitches` and `enableClinicalDownloads` be injected into the FlowController (from the builder that created it)? I think they both come from `AppConfiguration` or a similar "global" dependency.

Reply:
If they come globally, they would be read from `Current`.